src/lib.rs

```diff
@@ -1,36 +1,56 @@
-use std::ptr::null_mut;
-use std::os::raw::c_int;
+mod libxenevtchn;
+
+extern crate nix;
+extern crate xenevtchn_sys;
+use libxenevtchn::LibXenEvtchn;
+use log::debug;
 use std::io::Error;
+use std::os::raw::c_int;
+use std::ptr::null_mut;
+use xenevtchn_sys::{evtchn_port_t, xenevtchn_handle, xenevtchn_port_or_error_t};
 
 #[derive(Debug)]
 pub struct XenEventChannel {
-    handle: *mut xenevtchn_sys::xenevtchn_handle,
-    fd: c_int,
-    bind_port: c_int,
+    pub handle: *mut xenevtchn_handle,
+    pub bind_port: c_int,
+    libxenevtchn: LibXenEvtchn,
 }
 
 impl XenEventChannel {
+    pub fn new(domid: u32, evtchn_port: u32) -> Result<Self, Error> {
+        let handle = unsafe { xenevtchn_sys::xenevtchn_open(null_mut(), 0) };
+        if handle == null_mut() {
+            return Err(Error::last_os_error());
+        }
+        let bind_port =
+            unsafe { xenevtchn_sys::xenevtchn_bind_interdomain(handle, domid, evtchn_port) };
+        debug!("bind_port = {x}", x = bind_port);
+        if bind_port < 0 {
+            return Err(Error::last_os_error());
+        }
 
-    pub fn new(domid: u32, evtchn_port: u32) -> Result<Self,Error> {
-        let handle = unsafe {
-            xenevtchn_sys::xenevtchn_open(null_mut(), 0)
-        };
-        if handle == null_mut() { return Err(Error::last_os_error()); }
-
-        let fd = unsafe {
-            xenevtchn_sys::xenevtchn_fd(handle)
-        };
-        let bind_port = unsafe {
-            xenevtchn_sys::xenevtchn_bind_interdomain(handle, domid, evtchn_port)
-        };
-        if bind_port < 0 { return Err(Error::last_os_error()); }
+        let libxenevtchn = unsafe { LibXenEvtchn::new() };
 
         Ok(XenEventChannel {
             handle,
-            fd,
             bind_port,
+            libxenevtchn,
         })
     }
+
+    pub fn xenevtchn_pending(&self) -> Result<xenevtchn_port_or_error_t, Error> {
```

Comment:
return an Err if result is -1:
https://github.com/xen-project/xen/blob/437211cb696515ee5bd5dae0ab72866c9f382a33/tools/libs/evtchn/linux.c#L121